compiler/src/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -3585,159 +3498,830 @@ class ListOpConversion : public OpConversionPattern<SrcOpTy> {
           /*args=*/ArrayAttr{},
-          /*templateArgs=*/ArrayAttr{},
-          /*operands=*/ArrayRef<Value>(updatedOperands),
+          /*operands=*/ArrayRef<Value>(unwrappedOperands),
           /*typeConverter=*/*typeConverter);
 
-      rewriter.replaceOp(op, ArrayRef<Value>{});
+      if (failed(patchResults(op, adaptor, rewriter, unwrappedOperands,
+                              resultOperands))) {
+        return op.emitError() << "failed to patch operands";
+      }
+
+      if (failed(updateResults(op.getOperation(), rewriter, resultOperands))) {
+        return failure();
+      }
+
+      rewriter.eraseOp(op);
     } else {
       rewriter.replaceOpWithNewOp<emitc::CallOp>(
           /*op=*/op,
           /*type=*/op.getOperation()->getResultTypes(),
           /*callee=*/StringAttr::get(ctx, funcName),
           /*args=*/ArrayAttr{},
           /*templateArgs=*/ArrayAttr{},
-          /*operands=*/ArrayRef<Value>(updatedOperands));
+          /*operands=*/ArrayRef<Value>(unwrappedOperands));
+    }
+
+    return success();
+  }
+
+  // default behaviour
+  LogicalResult patchOperands(Operation *op, Adaptor adaptor,
+                              ConversionPatternRewriter &rewriter,
+                              SmallVector<Value, 4> &operands,
+                              SmallVector<Value, 4> &results) const {
+    if (!isa<IREE::VM::ListReserveOp, IREE::VM::ListResizeOp,
+             IREE::VM::ListSizeOp, IREE::VM::BufferLengthOp,
+             IREE::VM::BufferCopyOp>(op)) {
+      return op->emitError() << "either add this op to the allowed list to get "
+                                "the default behaviour or overload the "
+                                "`patchOperands` method.";
+    }
+    if (failable) {
+      if (failed(createDefaultOutOperands(op, rewriter, operands, results))) {
+        return failure();
+      }
+    }
+    return success();
+  }
+
+  LogicalResult patchResults(Operation *op, Adaptor adaptor,
+                             ConversionPatternRewriter &rewriter,
+                             SmallVector<Value, 4> &operands,
+                             SmallVector<Value, 4> &results) const {
+    if (!isa<IREE::VM::ListReserveOp, IREE::VM::ListResizeOp,
+             IREE::VM::ListSizeOp, IREE::VM::BufferLengthOp,
+             IREE::VM::BufferCopyOp, IREE::VM::BufferFillF32Op,
+             IREE::VM::BufferFillF64Op, IREE::VM::BufferFillI8Op,
+             IREE::VM::BufferFillI16Op, IREE::VM::BufferFillI32Op,
+             IREE::VM::BufferFillI64Op, IREE::VM::BufferStoreF32Op,
+             IREE::VM::BufferStoreF64Op, IREE::VM::BufferStoreI8Op,
+             IREE::VM::BufferStoreI16Op, IREE::VM::BufferStoreI32Op,
+             IREE::VM::BufferStoreI64Op>(op)) {
+      return op->emitError() << "either add this op to the allowed list to get "
+                                "the default behaviour or overload the "
+                                "`patchResults` method.";
+    }
```

Comment:
I'm having trouble following this patching stuff and this kind of switch list on concrete ops worries me - is there a good overview of what you're trying to do here and why it requires this kind of specialization? Changing the structure of these ops or adding a new one shouldn't require a bunch of deep changes in the emitc code and I'd like to make sure this isn't baking in structural assumptions. I'd rather see one function doing patching per op that's easy to find/fix/clone instead of a mix of generic things doing spooky stuff.

Reply:
Right, I moved this mapping from MLIR operands to C API operands into `ops.h`, so both compilation modes execute the same path.